### PR TITLE
Nested metric registries don't cause a sortedmap creation

### PR DESCRIPTION
**Pull request by j-baker**
This might seem like an unnecessary optimization, but profiling
internally implies this uses plenty of CPU and allocates gigabytes and
gigabytes and gigabytes (about 1.8GB of allocations to report metrics).

BLUF: With nested tags, our tooling spends all of its time recreating
nested sorted maps and rehashing, which is very expensive. That said,
breaking binary backcompat here is super expensive, as we know from past
experience.

This PR solves this issue. We no longer use the immutables generated
class, though we do use the builder. Instead, we use our own
implementation, which hashes consistently and uses two map
implementations to ensure that in the cases we see, equality, hash, and
extending with another tag are cheap.

### CODE REVIEW

**Comment by carterkozak** [comment]
Would you mind adding a representative JMH benchmark?

**Comment by j-baker** [comment]
That seems like a great idea.

**Comment by CRogers** [comment]
Chatted to @j-baker on slack, would be good to see a profile on this (possibly in prod) before we merge this as it's used by every product.

**Comment by carterkozak** [comment]
Could we do something with a TaggedMetricRegistry.forEach(BiConsumer<MetricName, Metric>) instead?

**Comment by j-baker** [comment]
I ended up doing both (and came to the forEach conclusion as well). Here are benchmark results before and after:

```
Before                                                                    Mode  Cnt        Score         Error   Units
NestedMetricsBenchmark.benchmarkForEach                                      avgt   10   949386.155 ±  290036.990   ns/op
NestedMetricsBenchmark.benchmarkForEach:·gc.alloc.rate.norm                  avgt   10  2122032.414 ±       0.121    B/op
NestedMetricsBenchmark.benchmarkGetMetrics                                   avgt   10  1627004.644 ±  361734.975   ns/op
NestedMetricsBenchmark.benchmarkGetMetrics:·gc.alloc.rate.norm               avgt   10  2800648.214 ±      12.257    B/op

After                                                                    Mode  Cnt        Score        Error   Units
NestedMetricsBenchmark.benchmarkForEach                                      avgt   10   214789.191 ±  66852.954   ns/op
NestedMetricsBenchmark.benchmarkForEach:·gc.alloc.rate.norm                  avgt   10   322032.094 ±      0.028    B/op
NestedMetricsBenchmark.benchmarkGetMetrics                                   avgt   10   557286.570 ±  78861.278   ns/op
NestedMetricsBenchmark.benchmarkGetMetrics:·gc.alloc.rate.norm               avgt   10   997784.244 ±      0.039    B/op
```

as you can see, even in the case where we only have a single pre-existing tag, this is a considerable improvement.

**Comment by j-baker** [comment]
This is also only to write 5k metrics - in practice we see many more.

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/PrehashedSortedMap.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/PrehashedSortedMap.java]: would suggest making immutability explicit in types

```suggestion
    PrehashedSortedMap(ImmutableSortedMap<K, V> delegate) {
```

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/PrehashedSortedMap.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/PrehashedSortedMap.java]: would suggest making immutability explicit in types

```suggestion
    private final ImmutableSortedMap<K, V> delegate;
```

**Comment by carterkozak** [comment]
I think we should update the benchmark into a root TaggedMetricRegistry with N tagged metrics, a child wrapped DropwizardMetricSet(MetricRegistry) with N metrics, and a child TaggedMetricRegistry with N metrics (representing SharedTaggedMetricRegistries.getSingleton()).
That should scale in a way that represents production systems.

I'd prefer to put all our eggs in the forEachMetric basket and avoid optimizations on getMetrics to reduce complexity. If we provide a simple MetricName implementation which doesn't pre-compute hashCode, and avoid precomputing hashcode on ExtraEntrySortedMap, we should have strictly less work.

Might be worth special casing adding a single tag to a MetricName with no tags, and adding a simple MetricName implementation around a safeName with no tags for DropwizardTaggedMetricSet.

File: tritium-registry/src/test/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMapTest.java
Comment (CRogers) [inline tritium-registry/src/test/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMapTest.java:35]: I've added this property based test as I wanted to make sure the behaviour of `ExtraEntrySortedMap` was correct but also not write a ton of example based tests. This checks that every method call on `ExtraEntrySortedMap` produces the same result as the same method call on an equivalent guava `ImmutableSortedMap`, for thousands of arbitrary base maps and extra values. It uses [jqwik](https://jqwik.net/docs/current/user-guide.html#creating-a-property).

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java]: can we assert non-null?
```suggestion
        this.base = Preconditions.checkNotNull(base, "base");
```

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java]: Understand suppressing `JdkObsolete` given `SortedMap`, but would like to not suppress the `NullableDereference ` and `NullAway` here if possible

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java]: seems like we assume `extraKey` and `extraValue` are non-null, can we assert?
```suggestion
        this.extraKey = Preconditions.checkNotNull(extraKey, "extraKey");
```

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java]: same as above
```suggestion
        this.extraValue = Preconditions.checkNotNull(extraValue, "extraValue");
```

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java:116]: if we don't ensure non-null `extraKey` at construction, this would later throw NPE

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java]: can avoid compiler warning with:

```suggestion
                return Iterators.transform(keySet().iterator(), key -> get(key));
```

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java:106]: is this true if the extra key is also in the base map?

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java:164]: is this true of extraKey is also in base map?

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java:179]: is this true of extraKey is also in base map?

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java:194]: is this true of extraKey is also in base map?

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java:212]: do we need to spread the hash more effectively given `String`'s poor hashCode?

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/PrehashedSortedMap.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/PrehashedSortedMap.java]: this assumes not only immutable map (i.e. can't add/remove from map), but also immutable entries such that you can't change the value for a given key, and that the keys/values themselves can't mutate or else the hashcode is no longer valid, and bad things happen

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/RealMetricName.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/RealMetricName.java:39]: just curious why djb2 hash algorithm?

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (CRogers) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java]: I wonder if we should check that `extraKey` does not exist in the base map. If this was an ordinary `ImmtuableSortedMap`, it would not allow you to do this, so this PR will open up the ability to "shadow" existing tags, which if people use may make it hard to revert to the previous behaviour, if needed.

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java]: Unfortunately ErrorProne is claiming that `ExtraEntrySortedMap.this` is null

Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java]: i can probably add some `@Nullables` to satisfy nullaway

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java:164]: oops, I meant to validate that one

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java:212]: `Map` defines the contract of `Map.hashCode` as sum of hashcode of entries (hashCode of entries is key.hashCode ^ value.hashCode)

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/PrehashedSortedMap.java
Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/PrehashedSortedMap.java]: yep - changed comment

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/RealMetricName.java
Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/RealMetricName.java:39]: it's the one in the immutables generated code

Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/RealMetricName.java:39]: copy-pasted :P

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java]: yeah, done

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/PrehashedSortedMap.java
Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/PrehashedSortedMap.java]: This would require casting in RealMetricName, which I'd like to avoid.

Comment (schlosna) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/PrehashedSortedMap.java]: not sure I follow that as the only caller of this constructor is `com.palantir.tritium.metrics.registry.RealMetricName#prehash` which explicitly passes in an `ImmutableSortedMap`

```
        return new PrehashedSortedMap<>(ImmutableSortedMap.copyOfSorted(map));
```

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java
Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExtraEntrySortedMap.java]: i solved all these i think

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/PrehashedSortedMap.java
Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/PrehashedSortedMap.java]: oops!

File: .baseline/checkstyle/checkstyle.xml
Comment (j-baker) [inline .baseline/checkstyle/checkstyle.xml:472]: as updated by baseline

Comment (schlosna) [reply .baseline/checkstyle/checkstyle.xml:472]: assume we will also pick this up next time the robots do their rounds in https://github.com/palantir/gradle-baseline/releases/tag/2.7.0 though I don't see a tritium excavator listed in https://github.com/palantir/gradle-baseline/pull/833

**Review by schlosna — APPROVED** [review]
LGTM 👍 , thanks James! Defer to you all when you want to merge & cut a release

**Pull request merged**
